### ISSUE TITLE
log() outputs nan in SoftDeterministicPolicyNetwork.sample()

### ISSUE THREAD

**Issue by syuntoku14** [cpnota/autonomous-learning-library#139]
Hi, cpnota! Your repo indeed helps me alot!

I reproduce your sac implementation and found a bug in SoftDeterministicPolicyNetwork.

The [torch.log(1 - action.pow(2) + 1e-6)](https://github.com/cpnota/autonomous-learning-library/blob/68d355a86dac95a1204ec327250a2e7e9494d92c/all/policies/soft_deterministic.py#L44) in line 44 can output nan if the tanh_scale is larger than 1.

According to appendix C in SAC paper, I think the _sample function should be:

```
    def _sample(self, normal):
        raw = normal.rsample()
        log_prob = normal.log_prob(raw)
        log_prob -= torch.log(1 - torch.tanh(raw).pow(2) + 1e-6)
        log_prob = log_prob.sum(1)
        action = self._squash(raw)
        return action, log_prob
```

Thank you!

**Comment by cpnota** [comment]
Hi again syuntoku14! I've been watching your il-rl repo, it looks awesome!

Thanks for this bug report, I'll take a look. I see your point, if `tanh_scale` was -2 and `raw` is a large negative number, we'll get the log of a negative number. Good catch! I think you are right that it's okay to not consider the effect of `_squash` on `log_prob`, because it's just a linear transformation, so the gradient will only be off by a positive factor.

**Comment by cpnota** [comment]
Closed by #140
